ports/espressif/common-hal/wifi/Radio.c

```diff
@@ -532,32 +532,59 @@ void common_hal_wifi_radio_set_ipv4_address_ap(wifi_radio_obj_t *self, mp_obj_t
     common_hal_wifi_radio_start_dhcp_server(self); // restart access point DHCP
 }
 
+static void ping_success_cb(esp_ping_handle_t hdl, void *args) {
+    wifi_radio_obj_t *self = (wifi_radio_obj_t *)args;
+    esp_ping_get_profile(hdl, ESP_PING_PROF_TIMEGAP, &self->ping_elapsed_time, sizeof(self->ping_elapsed_time));
+}
+
 mp_int_t common_hal_wifi_radio_ping(wifi_radio_obj_t *self, mp_obj_t ip_address, mp_float_t timeout) {
     esp_ping_config_t ping_config = ESP_PING_DEFAULT_CONFIG();
     ipaddress_ipaddress_to_esp_idf(ip_address, &ping_config.target_addr);
     ping_config.count = 1;
 
+    // We must fetch ping information using the callback mechanism, because the session storage is freed when
+    // the ping session is done, even before esp_ping_delete_session().
+    esp_ping_callbacks_t ping_callbacks = {
+        .on_ping_success = ping_success_cb,
+        .cb_args = (void *)self,
+    };
+
     size_t timeout_ms = timeout * 1000;
 
+    // ESP-IDF creates a task to do the ping session. It shuts down when done, but only after a one second delay.
+    // Calling common_hal_wifi_radio_ping() too fast will cause resource exhaustion.
     esp_ping_handle_t ping;
-    CHECK_ESP_RESULT(esp_ping_new_session(&ping_config, NULL, &ping));
+    esp_err_t ret;
+    for (size_t tries = 1; tries <= 5; tries++) {
+        ret = esp_ping_new_session(&ping_config, &ping_callbacks, &ping);
+        if (ret == ESP_OK) {
+            break;
+        }
+        // Wait for old task to go away and then try again.
+        // Empirical testing shows we have to wait at least two seconds, despite the task
+        // having a one-second timeout.
+        common_hal_time_delay_ms(2000);
+        if (mp_hal_is_interrupted()) {
+            return -1;
+        }
+    }
+    CHECK_ESP_RESULT(ret);
+
```

Comment:
(sorry if this appears twice; github seems to have eaten my comment)

I'm not super confident about this change.

 * I don't relish the idea of explaining to the user: yes, ping takes 2 seconds in time.monotonic but claims the ping response time is 0.1 seconds.
 * esp_ping_new_session might fail for other reasons (e.g., all sockets are in use) and now ping will wait 10 seconds before returning failure. before, it would fail prompt.y

Reply:
I've reworked this to wait a maximum of two seconds on failure, and also documented this.

I agree a slow call that returns a fast ping time is weird, but without rewriting the underlying ping implementation (which should probably be done, but I'm not going to do now), I think this is better than the current situation, which also requires explanation.

Reply:
@jepler Does the second commit alleviate your concerns?

@tannewt How is this PR for you?